crates/containerd-shim-wasm/src/sandbox/exec.rs

```diff
@@ -0,0 +1,252 @@
+use super::cgroups::{Cgroup, Version as CgroupVersion};
+use super::error::Error;
+use caps::{CapSet, Capability};
+use clone3::Clone3;
+use libc::siginfo_t;
+use log::debug;
+use log::info;
+use nix::sys::signal::SIGCHLD;
+use nix::sys::wait::{waitid, Id as WaitID, WaitPidFlag, WaitStatus};
+use nix::unistd::close;
+use std::os::raw::c_int as RawFD;
+use std::os::unix::io::AsRawFd;
+use std::ptr;
+
+#[derive(Clone)]
+pub struct PidFD {
+    fd: RawFD,
+}
+
+impl Drop for PidFD {
+    fn drop(&mut self) {
+        let _ = self.wait_no_hang(); // don't leave zombies
+        let _ = unsafe { libc::close(self.fd) };
+    }
+}
+unsafe fn pidfd_send_signal(pidfd: RawFD, sig: i32, info: *mut siginfo_t, flags: u32) -> i64 {
+    libc::syscall(libc::SYS_pidfd_send_signal, pidfd, sig, info, flags)
+}
+
+impl AsRawFd for PidFD {
+    fn as_raw_fd(&self) -> RawFD {
+        self.fd
+    }
+}
+
+#[derive(Debug)]
+pub enum Status {
+    Exited(ExitStatus),
+    Running,
+}
+
+impl From<Status> for ExitStatus {
+    fn from(status: Status) -> Self {
+        match status {
+            Status::Exited(status) => status,
+            Status::Running => panic!("cannot convert running process to exit status"),
+        }
+    }
+}
+
+fn wait(fd: RawFD, no_hang: bool) -> Result<Status, Error> {
+    let no_hang = if no_hang {
+        WaitPidFlag::WNOHANG
+    } else {
+        WaitPidFlag::empty()
+    };
+
+    let info = waitid(WaitID::PIDFd(fd), no_hang | WaitPidFlag::WEXITED)
+        .map_err(|e| std::io::Error::from(e))?;
+
+    match info {
+        WaitStatus::Exited(pid, status) => Ok(Status::Exited(ExitStatus {
+            pid: pid.as_raw() as u32,
+            status: status as u32,
+        })),
+        WaitStatus::Signaled(pid, sig, dumped) => {
+            info!("child {} killed by signal {}, dumped: {}", pid, sig, dumped);
+            Ok(Status::Exited(ExitStatus {
+                pid: pid.as_raw() as u32,
+                status: 128 + sig as u32,
+            }))
+        }
+        WaitStatus::StillAlive => Ok(Status::Running),
+        _ => Err(Error::Others(format!("unexpected wait status: {:?}", info))),
+    }
+}
+
+impl PidFD {
+    // wait for the process referred to by the pidfd to exit
+    //
+    // If you want more control over waiting you can use `as_raw_fd()` and call `waitid` yourself.
+    pub fn wait(&self) -> Result<ExitStatus, Error> {
+        let ws = wait(self.fd, false)?;
+        Ok(ws.into())
+    }
+
+    pub fn wait_no_hang(&self) -> Result<Status, Error> {
+        wait(self.fd, true)
+    }
+
+    // Send the specified signal to the process referred to by the pidfd.
+    pub fn kill(&self, sig: i32) -> Result<(), Error> {
```

Comment:
I think the usage here is safe since there is no way to mess up memory safety here except for the code inside of the `unsafe` block. There's nothing the caller can do to mess it up.